c/src/lib.rs

```diff
@@ -251,53 +283,107 @@ impl FoxgloveMcapOptions {
             .emit_attachment_indexes(self.emit_attachment_indexes)
             .emit_metadata_indexes(self.emit_metadata_indexes)
             .repeat_channels(self.repeat_channels)
-            .repeat_schemas(self.repeat_schemas)
+            .repeat_schemas(self.repeat_schemas))
     }
 }
 
 pub struct FoxgloveMcapWriter(Option<foxglove::McapWriterHandle<BufWriter<File>>>);
 
 /// Create or open an MCAP file for writing. Must later be freed with `foxglove_mcap_free`.
 ///
+/// If an error occurs, returns null and error will be set with the error details.
+/// Remember to call `foxglove_error_free` afterwards.
+/// If error is null, no error details are returned.
+///
 /// # Safety
-/// `path`, `profile`, and `library` must be valid UTF8.
+/// `path` and `profile` must be valid UTF8.
 #[unsafe(no_mangle)]
 #[must_use]
 pub unsafe extern "C" fn foxglove_mcap_open(
     options: &FoxgloveMcapOptions,
+    error: *mut FoxgloveError,
 ) -> *mut FoxgloveMcapWriter {
+    unsafe {
+        match do_foxglove_mcap_open(options) {
+            Ok(writer) => Box::into_raw(writer),
+            Err(err) => {
+                set_error(error, err);
+                std::ptr::null_mut()
+            }
+        }
+    }
+}
+
+unsafe fn do_foxglove_mcap_open(
+    options: &FoxgloveMcapOptions,
+) -> Result<Box<FoxgloveMcapWriter>, foxglove::FoxgloveError> {
     let path = std::str::from_utf8(unsafe {
         std::slice::from_raw_parts(options.path as *const u8, options.path_len)
     })
-    .expect("path is invalid");
+    .map_err(|e| foxglove::FoxgloveError::ValueError(format!("path is invalid: {}", e)))?;
 
     // Safety: this is safe if the options struct contains valid strings
-    let mcap_options = unsafe { options.to_write_options() };
+    let mcap_options = unsafe { options.to_write_options() }?;
+
+    println!("create, truncate: {}, {}", options.create, options.truncate);
 
-    let file = File::options()
-        .write(true)
-        .create(options.create)
-        .truncate(options.truncate)
+    let mut file_options = File::options();
+    file_options.write(true);
+    if options.create {
+        if options.truncate {
+            file_options.create(true).truncate(true);
+        } else {
+            file_options.create_new(true);
+        }
+    } else if options.truncate {
+        file_options.truncate(true);
+    } else {
+        // Append doesn't make sense with mcap
+        return Err(foxglove::FoxgloveError::IoError(std::io::Error::new(
```

Comment:
How much benefit do you see to supporting create=false & truncate=true (requiring that a file already exist)? The downside is that these options open the possibility for this unsupported case.

If it's a net benefit, should python have a similar interface? Right now, it only provides the equivalent of a truncate option.

Reply:
That's a good point, I don't think that combination makes sense.

I think just a truncate option is enough, which just describes whether we error or truncate an existing file.

Create can be assumed to always be true.